server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

```diff
@@ -0,0 +1,167 @@
+            return Decision.single(Decision.Type.YES, NAME, "Decider allows replicas move to index nodes.");
+        }
+
+        if (node.node().getRoles().contains(SEARCH_ROLE) && shardRouting.primary()) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider allows primaries move to search nodes.");
+        }
+
+        final ProjectId projectId = allocation.getClusterState().metadata().projectFor(index).id();
+        final Set<DiscoveryNode> eligibleNodes = new HashSet<>();
+        int totalShards = 0;
+        String nomenclature = EMPTY;
+
+        if (node.node().getRoles().contains(INDEX_ROLE)) {
+            collectEligibleNodes(allocation, eligibleNodes, INDEX_ROLE);
+            // Primary shards only.
+            totalShards = allocation.getClusterState().routingTable(projectId).index(index).size();
+            nomenclature = "primary shards";
+        } else if (node.node().getRoles().contains(SEARCH_ROLE)) {
+            collectEligibleNodes(allocation, eligibleNodes, SEARCH_ROLE);
+            // Replicas only.
+            final IndexMetadata indexMetadata = allocation.getClusterState().metadata().getProject(projectId).index(index);
+            totalShards = indexMetadata.getNumberOfShards() * indexMetadata.getNumberOfReplicas();
+            nomenclature = "replicas";
+        }
+
+        assert eligibleNodes.isEmpty() == false;
+        assert totalShards > 0;
+
+        final double idealAllocation = Math.ceil((double) totalShards / eligibleNodes.size());
+        final int threshold = (int) Math.ceil(idealAllocation * indexBalanceConstraintSettings.getLoadSkewTolerance());
+        final int currentAllocation = node.numberOfOwningShardsForIndex(index);
+
+        if (currentAllocation >= threshold) {
+            String explanation = Strings.format(
+                """
+                    For index [%s] with [%d] %s, Node [%s] is expected to hold [%.0f] %s for index [%s], based on the total of [%d]
+                    nodes available. The configured load skew tolerance is [%.2f], which yields an allocation threshold of
+                    Math.ceil([%.0f] × [%.2f]) = [%d] %s. Currently, node [%s] is assigned [%d] %s of index [%s]. Therefore,
+                    assigning additional %s is not preferred.
+                    """,
+                index,
+                totalShards,
+                nomenclature,
+                node.nodeId(),
+                idealAllocation,
+                nomenclature,
+                index,
+                eligibleNodes.size(),
+                indexBalanceConstraintSettings.getLoadSkewTolerance(),
+                idealAllocation,
+                indexBalanceConstraintSettings.getLoadSkewTolerance(),
+                threshold,
+                nomenclature,
+                node.nodeId(),
+                currentAllocation,
+                nomenclature,
+                index,
+                nomenclature
```

Comment:
There is a lot of repeated information in this string. How about tightening it up to something like
```
"""
    There are [5] eligible nodes in the [index] tier for assignment of [4] shards in index [idx]. Ideally no more than [2]
    shard would be assigned per node (the index balance skew setting is [1.5]). This node is already assigned [3] shards of
    the index.
"""
```
Deduplicating the parameter list to
```
    eligibleNodes.size(),
    nomenclature,
    totalShards,
    index,
    idealAllocation,
    indexBalanceConstraintSettings.getLoadSkewTolerance(),
    currentAllocation
```